presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java

```diff
@@ -933,6 +933,29 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
         return Optional.empty();
     }
 
+    /**
+     * Attempt to push down the TopN into the table scan.
+     * <p>
+     * Connectors can indicate whether they don't support topN pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}. Connectors should expect this method may be called multiple times.
+     * </p>
+     * <b>Note</b>: it's critical for connectors to return {@link Optional#empty()} if calling this method has no effect for that
+     * invocation, even if the connector generally supports topN pushdown. Doing otherwise can cause the optimizer
+     * to loop indefinitely.
+     * <p>
+     * If the connector can handle TopN Pushdown it should return a new table handle which will replace the existing
+     * table handle in the TableScan.
+     */
+    default Optional<ConnectorTableHandle> applyTopN(
```

Comment:
I think we'll want to return a structure similar to `applyLimit`, where it indicates whether the limit is guaranteed by the connector. In the case of a distributed connector implementation, it may not be able to perform a global top N, but it could still benefit from pushing down the operation, for instance, to each individual storage shard.

Reply:
Added TopNApplicationResult which is basically same as LimitApplicationResult for now and added a test case that tests that when limitGuaranteed is set to false we create a plan with TopNPartial stage with updated connector table handle reference.

Reply:
@martint can you review this again when you find some time?